src/mongo/db/pipeline/expression_test.cpp

```diff
@@ -2894,6 +2894,92 @@ TEST(ExpressionObjectOptimizations,
 
 }  // namespace Object
 
+TEST(ExpressionFilter, ExpressionFilterWithASetLimitShouldReturnAArrayNoGreaterThanTheLimit) {
+    intrusive_ptr<ExpressionContextForTest> expCtx(new ExpressionContextForTest());
+    VariablesParseState vps = expCtx->variablesParseState;
+    auto filterSpec = BSON(
+        "$filter" << BSON("input" << BSON_ARRAY(1 << 2 << 3 << 4 << 5 << 6 << 7 << 8 << 9) << "as"
+                                  << "arr"
+                                  << "cond" << BSON("$gt" << BSON_ARRAY("$$arr" << 3))));
+
+
+    auto expFilter = ExpressionFilter::parse(expCtx, filterSpec.firstElement(), vps);
+    dynamic_cast<ExpressionFilter*>(expFilter.get())->setLimit(1);
+    auto oneElemArray = expFilter->evaluate(Document());
+    ASSERT_TRUE(oneElemArray.getArray().size() == 1);
+    ASSERT_VALUE_EQ(oneElemArray, Value(BSON_ARRAY(4)));
+
+    dynamic_cast<ExpressionFilter*>(expFilter.get())->setLimit(2);
+    auto twoElemArray = expFilter->evaluate(Document());
+    ASSERT_TRUE(twoElemArray.getArray().size() == 2);
+    ASSERT_VALUE_EQ(twoElemArray, Value(BSON_ARRAY(4 << 5)));
+
+    dynamic_cast<ExpressionFilter*>(expFilter.get())->setLimit(5);
+    auto fiveElemArray = expFilter->evaluate(Document());
+    ASSERT_TRUE(fiveElemArray.getArray().size() == 5);
+    ASSERT_VALUE_EQ(fiveElemArray, Value(BSON_ARRAY(4 << 5 << 6 << 7 << 8)));
+    // Filter runs out of elements before limit is reached
+    dynamic_cast<ExpressionFilter*>(expFilter.get())->setLimit(10);
+    auto sixElemArray = expFilter->evaluate(Document());
+    ASSERT_TRUE(sixElemArray.getArray().size() == 6);
+    ASSERT_VALUE_EQ(sixElemArray, Value(BSON_ARRAY(4 << 5 << 6 << 7 << 8 << 9)));
+}
+TEST(ExpressionArrayElemAt, ArrayElemAtWithFilterShouldEvaluateCorrectly) {
+    intrusive_ptr<ExpressionContextForTest> expCtx(new ExpressionContextForTest());
+    VariablesParseState vps = expCtx->variablesParseState;
+    // Returns an array with all values greater than 3.
+    auto filterSpec = BSON(
+        "$filter" << BSON("input" << BSON_ARRAY(1 << 2 << 3 << 4 << 5 << 6 << 7 << 8 << 9) << "as"
+                                  << "arr"
+                                  << "cond" << BSON("$gt" << BSON_ARRAY("$$arr" << 3))));
+
+    auto arrayElemAtSpec = BSON("$arrayElemAt" << BSON_ARRAY(filterSpec << 2));
+
+    auto expArrayElemAt = ExpressionArrayElemAt::parse(expCtx, arrayElemAtSpec.firstElement(), vps);
+    auto xpArrayElemAt = dynamic_cast<ExpressionArrayElemAt*>(expArrayElemAt.get())->optimize();
+    auto val = xpArrayElemAt->evaluate(Document());
+    ASSERT_VALUE_EQ(val, Value(6));
+
+    auto eA = ExpressionArrayElemAt::parse(expCtx, BSON("$arrayElemAt" << BSON_ARRAY(BSON_ARRAY(1 << 2 <<3 << 4 <<5) << 1)).firstElement(), vps);
```

Comment:
You can re-use the variable names from before, I think that'd be clearer than two-letter variable names.

Reply:
Same with below. Alternatively, you could separate each one into it's own test.